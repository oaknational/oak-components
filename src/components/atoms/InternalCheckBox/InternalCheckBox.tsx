import React, { forwardRef, useRef } from "react";
import styled, { css } from "styled-components";

import { OakBorderRadiusToken, OakCombinedColorToken } from "@/styles";
import { parseBorderRadius } from "@/styles/helpers/parseBorderRadius";
import { parseColor } from "@/styles/helpers/parseColor";
import { parseDropShadow } from "@/styles/helpers/parseDropShadow";
import { BorderStyleProps, borderStyle } from "@/styles/utils/borderStyle";
import { ColorStyleProps, colorStyle } from "@/styles/utils/colorStyle";
import { SizeStyleProps, sizeStyle } from "@/styles/utils/sizeStyle";
import { SpacingStyleProps, spacingStyle } from "@/styles/utils/spacingStyle";

/**
 *
 * These components can be used with InternalCheckBoxWrapper which allows for customisable icons
 *
 * Several flavours of checkbox are created here:
 *  - Default
 *  - Hover decorations
 *  - Focus decorations
 *  - Hover + Focus decorations
 *
 * As they are styled components they can be further customised in implementation. Alternatively additional
 * components can be created here.
 *
 */

export type BaseCheckBoxProps = {
  id: string;
  disabled?: boolean;
  value: string;
  /**
   * Uncontrolled checked state
   */
  defaultChecked?: boolean;
  /**
   * Controlled checked state
   */
  checked?: boolean;
  onHovered?: (value: string, id: string, duration: number) => void;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
  "data-testid"?: string;
};

const BaseCheckBox = forwardRef(
  (props: BaseCheckBoxProps, ref?: React.Ref<HTMLInputElement>) => {
    const { onHovered, ...rest } = props;

    const hoverStart = useRef(Date.now());

    const handleMouseEnter = () => {
      hoverStart.current = Date.now();
    };

    const handleMouseLeave = () => {
      const delta = Date.now() - hoverStart.current;
      if (onHovered) {
        onHovered(props.value, props.id, delta);
      }
    };

    return (
      <input
        ref={ref}
        type="checkbox"
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        {...rest}
        name={props.id}
      />
    );
  },
);

type StyledBaseCheckBoxProps = BaseCheckBoxProps &
  ColorStyleProps &
  SpacingStyleProps &
  BorderStyleProps &
  SizeStyleProps & {
    $checkedBackground?: OakCombinedColorToken | null;
  };

type HoverBaseCheckBoxProps = {
  $hoverBorderRadius: OakBorderRadiusToken;
};

/**
 *
 * These components can be used with InternalCheckBoxWrapper which allows for customisable icons
 *
 * Several flavours of checkbox are created here:
 *  - Default
 *  - Hover decorations
 *  - Focus decorations
 *  - Hover + Focus decorations
 *
 * NB. Hover decorations must be wrapped in a box with position relative to allow for the hover effect to work
 *
 * As they are styled components they can be further customised in implementation. Alternatively additional
 * components can be created here.
 *
 *
 * ## Events
 * The following callbacks are available for tracking focus events:
 *
 * ### onChange
 * onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
 *
 * ### onFocus
 *   onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
 *
 * ### onBlur
 *    onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
 *
 * ### onHovered
 *  `onHovered?: (id, value, duration: number) => void;`<br>
 *  called after a mouseEnter and mouseLeave event has happened
 *
 *
 */
export const InternalCheckBox = styled(BaseCheckBox)<StyledBaseCheckBoxProps>`
  /* removing default appearance */
  -webkit-appearance: none;
  appearance: none;
  cursor: pointer;
  margin: 0;
  outline: none;
  ${borderStyle}
  ${colorStyle}
  ${spacingStyle}
  ${sizeStyle}

  &:checked {
    ${(props) => css`
      background: ${parseColor(props.$checkedBackground)};
    `};
  }

  &:disabled {
    pointer-events: none;
  }
`;

InternalCheckBox.defaultProps = {
  $borderRadius: "border-radius-xs",
  $ba: "border-solid-m",
  $borderColor: "text-primary",
  $checkedBackground: "text-primary",
};

export const InternalCheckBoxHover = styled(InternalCheckBox)<
  StyledBaseCheckBoxProps & HoverBaseCheckBoxProps
>`
  /* @media wrapper is required to prevent hover effect on iOS Safari */

  @media (hover: hover) {
    &:hover:not(&:checked):not(&:disabled)::after {
      content: "";
      position: absolute;
      top: 50%;
      left: 50%;
      width: 60%;
      height: 60%;
      transform: translate(-50%, -50%);
      border-radius: ${(props) => css`
        ${parseBorderRadius(props.$hoverBorderRadius)}
      `};
      background: ${(props) => css`
        ${parseColor(props.$checkedBackground)}
      `};
    }
  }
`;

InternalCheckBoxHover.defaultProps = {
  $hoverBorderRadius: "border-radius-xs",
};

const focusStyle = css`
  &:focus-visible {
    box-shadow: ${parseDropShadow("drop-shadow-centered-lemon")};
  }
`;

export const InternalCheckBoxFocus = styled(
  InternalCheckBox,
)<StyledBaseCheckBoxProps>`
  ${focusStyle}
`;

export const InternalCheckBoxHoverFocus = styled(InternalCheckBoxHover)<
  StyledBaseCheckBoxProps & HoverBaseCheckBoxProps
>`
  ${focusStyle}
`;
