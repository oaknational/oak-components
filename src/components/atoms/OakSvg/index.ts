export { OakSvg } from "./OakSvg";
export type { OakSvgProps, OakSvgNames } from "./OakSvg";
