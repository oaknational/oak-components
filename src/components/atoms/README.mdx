# Atoms

## Rules for components in this folder

- Can import other atoms
- Imports from molecules and organisms are forbidden!
- Generic
- Unstyled
- Avoid embedded logic and state
- No sub-components
  (eg. Box)
