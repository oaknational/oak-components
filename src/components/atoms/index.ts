export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
export * from "./OakIcon";
export * from "./OakGrid";
export * from "./OakGridArea";
export * from "./OakThemeProvider";
export * from "./OakMaxWidth";
export * from "./OakCloudinaryImage";
export * from "./OakKbd";
export * from "./OakGlobalStyle";
export * from "./OakAnchorTarget";
export * from "./OakFieldSet";
