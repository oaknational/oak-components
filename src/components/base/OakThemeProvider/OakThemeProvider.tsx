import React from "react";
import { ThemeProvider } from "styled-components";

import { OakTheme } from "@/styles";

export type OakThemeProviderProps = {
  theme: OakTheme;
  children: React.ReactNode;
};

export const OakThemeProvider = ({
  theme,
  children,
}: OakThemeProviderProps) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
