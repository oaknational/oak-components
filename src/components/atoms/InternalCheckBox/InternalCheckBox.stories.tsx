import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import {
  InternalCheckBox,
  InternalCheckBoxHover,
  InternalCheckBoxFocus,
  InternalCheckBoxHoverFocus,
} from "./InternalCheckBox";

import { OakBox } from "@/components/atoms/OakBox";
import { colorArgTypes } from "@/storybook-helpers/colorStyleHelpers";
import { spacingArgTypes } from "@/storybook-helpers/spacingStyleHelpers";
import { borderArgTypes } from "@/storybook-helpers/borderStyleHelpers";
import { sizeArgTypes } from "@/storybook-helpers/sizeStyleHelpers";

const meta: Meta<typeof InternalCheckBox> = {
  component: InternalCheckBox,
  tags: ["autodocs"],
  title: "components/atoms/InternalCheckBox",
  argTypes: {
    ...sizeArgTypes,
    ...borderArgTypes,
    ...colorArgTypes,
    ...spacingArgTypes,
    disabled: {
      control: "boolean",
    },
    $checkedBackground: colorArgTypes.$color,
  },
  parameters: {
    controls: {
      include: [
        "disabled",
        "defaultChecked",
        "$width",
        "$height",
        "$ba",
        "$borderRadius",
        "$borderColor",
        "$hoverBorderRadius",
        "$checkedBackground",
      ],
    },
  },
};
export default meta;

export const Default: StoryObj<typeof InternalCheckBox> = {
  render: (args) => <InternalCheckBox {...args} />,
  args: {
    id: "checkbox-test-default-1",
    value: "a test value",
    $width: "all-spacing-6",
    $height: "all-spacing-6",
  },
};

export const Hover: StoryObj<typeof InternalCheckBoxHover> = {
  render: (args) => (
    <OakBox $position="relative" $width={args.$width} $height={args.$height}>
      <InternalCheckBoxHover {...args} />
    </OakBox>
  ),
  args: {
    id: "checkbox-test-default-2",
    value: "a test value",
    $width: "all-spacing-6",
    $height: "all-spacing-6",
  },
};

export const Focus: StoryObj<typeof InternalCheckBoxFocus> = {
  render: (args) => (
    <OakBox $position="relative" $width={args.$width} $height={args.$height}>
      <InternalCheckBoxFocus {...args} />
    </OakBox>
  ),
  args: {
    id: "checkbox-test-default-3",
    value: "a test value",
    $width: "all-spacing-6",
    $height: "all-spacing-6",
  },
};

export const HoverFocus: StoryObj<typeof InternalCheckBoxHoverFocus> = {
  render: (args) => (
    <OakBox $position="relative" $width={args.$width} $height={args.$height}>
      <InternalCheckBoxHoverFocus {...args} />
    </OakBox>
  ),
  args: {
    id: "checkbox-test-default-4",
    value: "a test value",
    $width: "all-spacing-6",
    $height: "all-spacing-6",
  },
};
