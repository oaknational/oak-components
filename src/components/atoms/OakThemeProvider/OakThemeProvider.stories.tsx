import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakThemeProvider } from "./OakThemeProvider";

import { OakFlex } from "@/components/atoms/OakFlex";
import { oakDefaultTheme } from "@/styles";

const meta: Meta<typeof OakThemeProvider> = {
  component: OakThemeProvider,
  tags: ["autodocs"],
  title: "components/atoms/OakThemeProvider",
  argTypes: {},
  parameters: {
    controls: {
      include: [],
      sort: "none",
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakThemeProvider>;

export const DefaultOakThemeProvider: Story = () => (
  <OakThemeProvider theme={oakDefaultTheme}>
    <OakFlex
      $background="bg-decorative1-main"
      $color="text-inverted"
      $ba="border-solid-m"
      $pa="spacing-24"
      $borderRadius="border-radius-l"
      $borderColor="border-primary"
    >
      Get a head-start on your lesson planning using quality-checked resources
      you can download and adapt for free.
    </OakFlex>
  </OakThemeProvider>
);

DefaultOakThemeProvider.args = {};
