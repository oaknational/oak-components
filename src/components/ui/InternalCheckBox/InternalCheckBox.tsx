import React from "react";
import styled from "styled-components";

import {
  SubBaseCheckBoxProps,
  SubStyledCheckBox,
  SubStyledCheckBoxDecor,
} from "./SubStyledCheckBox";

import { OakBox, OakBoxProps, OakIcon } from "@/components/base";
import {
  OakAllSpacingToken,
  OakBorderRadiusToken,
  OakBorderWidthToken,
  OakCombinedColorToken,
  OakInnerPaddingToken,
} from "@/styles";
import { ResponsiveValues } from "@/styles/utils/responsiveStyle";

const StyledIconContainer = styled(OakBox)<
  OakBoxProps & { disabled?: boolean }
>`
  pointer-events: none;
  opacity: 0;

  input:checked + & {
    opacity: 1;
  }
`;

export type InternalCheckBoxProps = SubBaseCheckBoxProps & {
  size?: ResponsiveValues<OakAllSpacingToken>;
  decor?: SubStyledCheckBoxDecor;
  border?: ResponsiveValues<OakBorderWidthToken>;
  borderRadius?: ResponsiveValues<OakBorderRadiusToken>;
  borderColor?: OakCombinedColorToken;
  checkedBackground?: OakCombinedColorToken | null;
  checkedIcon?: React.JSX.Element;
  hoverBorderRadius?: OakBorderRadiusToken;
  iconPadding?: OakInnerPaddingToken;
};

const CheckBox = (props: InternalCheckBoxProps) => {
  const {
    size = "all-spacing-6",
    border = "border-solid-m",
    borderRadius = "border-radius-xs",
    borderColor = "text-primary",
    checkedBackground = "text-primary",
    iconPadding = "inner-padding-none",
    hoverBorderRadius = "border-radius-xs",
    ...rest
  } = props;

  const {
    checkedIcon = (
      <OakIcon
        iconName="tick"
        $width={"100%"}
        $height={"100%"}
        $colorFilter={"white"}
      />
    ),
  } = props;

  return (
    <OakBox $position="relative" $width={size} $height={size}>
      <SubStyledCheckBox
        $width={size}
        $height={size}
        $ba={border}
        $borderRadius={borderRadius}
        $borderColor={borderColor}
        $checkedBackground={checkedBackground}
        $hoverBorderRadius={hoverBorderRadius}
        {...rest}
      />

      <StyledIconContainer
        $position={"absolute"}
        $top={"all-spacing-0"}
        $left={"all-spacing-0"}
        $pa={iconPadding}
        $width={size}
        $height={size}
      >
        {checkedIcon}
      </StyledIconContainer>
    </OakBox>
  );
};

export const InternalCheckBox = styled(CheckBox)``;
