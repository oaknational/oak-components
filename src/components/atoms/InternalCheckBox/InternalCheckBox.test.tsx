import React, { createRef } from "react";
import "@testing-library/jest-dom";
import { create } from "react-test-renderer";
import { fireEvent } from "@testing-library/react";

import { InternalCheckBox } from "./InternalCheckBox";

import renderWithTheme from "@/test-helpers/renderWithTheme";
import { OakThemeProvider } from "@/components/atoms/OakThemeProvider";
import { oakDefaultTheme } from "@/styles";

describe("InternalCheckBox", () => {
  it("renders a checkbox", () => {
    const { getByRole } = renderWithTheme(
      <InternalCheckBox
        id="checkbox-1"
        value="Option 1"
        data-testid="test-1"
      />,
    );
    expect(getByRole("checkbox")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = create(
      <OakThemeProvider theme={oakDefaultTheme}>
        <InternalCheckBox id="checkbox-1" value="Option 1" />
      </OakThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("has a role of checkbox", () => {
    const { getByRole } = renderWithTheme(
      <InternalCheckBox id="checkbox-1" value="Option 1" />,
    );

    expect(getByRole("checkbox")).toBeInTheDocument();
  });

  it("has a name attribute of value id", () => {
    const { getByRole } = renderWithTheme(
      <InternalCheckBox id="checkbox-1" value="Option 1" />,
    );

    expect(getByRole("checkbox")).toHaveAttribute("name", "checkbox-1");
  });

  it("can be checked and unchecked through clicking", () => {
    const { getByRole } = renderWithTheme(
      <InternalCheckBox id="checkbox-1" value="Option 1" />,
    );
    getByRole("checkbox").click();
    expect(getByRole("checkbox")).toBeChecked();
    getByRole("checkbox").click();
    expect(getByRole("checkbox")).not.toBeChecked();
  });

  it("can be clicked through its ref", () => {
    const ref = createRef<HTMLInputElement>();
    const { getByRole } = renderWithTheme(
      <InternalCheckBox id="checkbox-1" value="Option 1" ref={ref} />,
    );
    ref.current?.click();
    expect(getByRole("checkbox")).toBeChecked();
    ref.current?.click();
    expect(getByRole("checkbox")).not.toBeChecked();
  });

  it("is uncheckable when disabled", () => {
    const { getByRole } = renderWithTheme(
      <InternalCheckBox id="checkbox-1" value="Option 1" disabled />,
    );
    getByRole("checkbox").click();
    expect(getByRole("checkbox")).not.toBeChecked();
  });

  it("calls onChange method when checked and unchecked", () => {
    const onChange = jest.fn();
    const { getByRole } = renderWithTheme(
      <InternalCheckBox id="checkbox-1" value="Option 1" onChange={onChange} />,
    );
    getByRole("checkbox").click();
    getByRole("checkbox").click();
    expect(onChange).toHaveBeenCalledTimes(2);
  });

  it("calls onFocus and onBlur with focus and blur", () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const { getByRole } = renderWithTheme(
      <InternalCheckBox
        id="checkbox-1"
        value="Option 1"
        onFocus={onFocus}
        onBlur={onBlur}
      />,
    );
    getByRole("checkbox").focus();
    getByRole("checkbox").blur();
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onBlur).toHaveBeenCalledTimes(1);
  });

  it("calls onHovered with mouseenter and mouseleave", () => {
    const onHovered = jest.fn();
    const { getByRole } = renderWithTheme(
      <InternalCheckBox
        id="checkbox-1"
        value="Option 1"
        onHovered={onHovered}
      />,
    );
    fireEvent.mouseEnter(getByRole("checkbox"));
    fireEvent.mouseLeave(getByRole("checkbox"));
    expect(onHovered).toHaveBeenCalledTimes(1);
  });
});
