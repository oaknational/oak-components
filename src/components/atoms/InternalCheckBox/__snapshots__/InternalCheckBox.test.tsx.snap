// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalCheckBox matches snapshot 1`] = `
.c0 {
  -webkit-appearance: none;
  appearance: none;
  cursor: pointer;
  margin: 0;
  outline: none;
  border: 0.125rem solid;
  border-color: #222222;
  border-radius: 0.125rem;
}

.c0:checked {
  background: #222222;
}

.c0:disabled {
  pointer-events: none;
}

@media (hover: hover) {
  .c0:hover:not(:disabled) {
    border-color: #222222;
  }
}

<input
  className="c0"
  id="checkbox-1"
  name="checkbox-1"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  type="checkbox"
  value="Option 1"
/>
`;
